lib/os/onoff.c

```diff
@@ -0,0 +1,522 @@
+		cli->async.callback.handler(srv, cli,
+					    cli->async.callback.user_data, res);
+		break;
+#ifdef CONFIG_POLL
+	case ONOFF_CLIENT_NOTIFY_SIGNAL:
+		k_poll_signal_raise(cli->async.signal, res);
+		break;
+#endif /* CONFIG_POLL */
+	default:
+		__ASSERT_NO_MSG(false);
+	}
+}
+
+static void notify_all(struct onoff_service *srv,
+		       sys_slist_t *list,
+		       int res)
+{
+	while (!sys_slist_is_empty(list)) {
+		sys_snode_t *node = sys_slist_get_not_empty(list);
+		struct onoff_client *cli =
+			CONTAINER_OF(node,
+				     struct onoff_client,
+				     node);
+
+		notify_one(srv, cli, res);
+	}
+}
+
+static void onoff_stop_notify(struct onoff_service *srv,
+			      int res);
+
+static void onoff_start_notify(struct onoff_service *srv,
+			       int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+
+	/* Can't have a queued releaser during start */
+	__ASSERT_NO_MSG(srv->releaser == NULL);
+
+	/* If the start failed log an error and leave the rest of the
+	 * state in place for diagnostics.
+	 *
+	 * If the start succeeded record a reference for all clients
+	 * and set the state to ON.  There must be at least one client
+	 * left to receive the result.
+	 *
+	 * In either case reset the client queue and notify all
+	 * clients of operation completion.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+		srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+	} else {
+		sys_snode_t *node;
+		unsigned int refs = 0U;
+
+		set_service_state(srv, SERVICE_STATE_ON);
+
+		SYS_SLIST_FOR_EACH_NODE(&clients, node) {
+			refs += 1U;
+		}
+
+		/* Update the reference count, or fail if the count
+		 * would overflow.
+		 */
+		if (srv->refs > (SERVICE_REFS_MAX - refs)) {
+			srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+		} else {
+			srv->refs += refs;
+		}
+		__ASSERT_NO_MSG(srv->refs > 0U);
+	}
+
+	sys_slist_init(&srv->clients);
+
+	k_spin_unlock(&srv->lock, key);
+
+	notify_all(srv, &clients, res);
+}
+
+int onoff_request(struct onoff_service *srv,
+		  struct onoff_client *cli)
+{
+	bool add_client = false;        /* add client to pending list */
+	bool start = false;             /* invoke start transition */
+	bool notify = false;            /* do client notification */
+	int rv = validate_args(srv, cli);
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) != 0) {
+		rv = -EIO;
+		goto out;
+	}
+
+	/* Reject if this would overflow the reference count. */
+	if (srv->refs == SERVICE_REFS_MAX) {
+		rv = -EAGAIN;
+		goto out;
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_TO_OFF:
+		/* Queue to start after release */
+		__ASSERT_NO_MSG(srv->releaser != NULL);
+		add_client = true;
+		rv = 3;
+		break;
+	case SERVICE_STATE_OFF:
+		/* Reject if in a non-thread context and start could
+		 * wait.
+		 */
+		if ((k_is_in_isr() || k_is_pre_kernel())
```

Comment:
this check is done twice. Would be good to extract to a function (`is_in_sleepable_context()`). What about context which has interrupts locked? See https://github.com/zephyrproject-rtos/zephyr/blob/8617af383d7173e04d2cce4739a687940faa7e16/subsys/logging/log_msg.c#L71

Reply:
If anything gets extracted it should be `k_is_thread_context()` defined as `!(k_is_in_isr() || k_is_pre_kernel())`.  Which probably be a helper in `kernel.h`.

I don't think checking for locked interrupts is relevant.  AFAICT only the logging infrastructure is doing this, dunno why.

Reply:
> I don't think checking for locked interrupts is relevant. 

Right, in case of logging we want to make sure that we block only in preemptable context. Even though, there is no kernel function, there could be local function that would encapsulate that.

Reply:
It could be abstracted locally, but I don't see that as improving the code.  The current idiom should be good until `k_is_in_thread_context()` is warranted.